### PR TITLE
Fix spinner cheese by accounting for spin directionality

### LINKED ISSUE #8224: Spinners can be cheesed
**Describe the bug:** 

By tapping extremely rapidly with two fingers with the circle of the spinner at the middle. You can go up to 430 RPM.

**Screenshots or videos showing encountered issue:** 

[Video](https://player.vimeo.com/video/396941387) 

**osu!lazer version:** 

2020.311.0 

**Logs:**
 Android

### PR DESCRIPTION
**Pull request by peppy**
- Supersedes https://github.com/ppy/osu/pull/24999
- Closes #8224

I've simplified things in this PR. Hopefully I've left enough inline commenting to explain the reasoning behind things.

Have a read through this and see what you think. I'm up for explaining rationale over voice if it helps, but my summary thoughts are (mostly @smoogipoo as we have had countless of hours of discussion and both understand the caveats we encountered in previous implementations):

Without proper important frame accounting of spinners, this is as-good-as-or-better-than master/stable. Given two frames, the maximum angle we can accurately calculate is 180 degrees. So the goal going forward should be to, for the sake of replay accuracy, factor this in to important frames.

That is out of scope of this change, but is a consideration of this implementation. It simplifies things greatly in both code and thought process.

I have not given due diligence to tests. These are taken from #24999. The only changes/additions I made to tests are in 9011170c3e4df5ccf26ff762da2972c7f4d0bd0c and 80ac0c46d9895e40171445552551f76ff7f2b118.

### CODE REVIEW

**Review by bdach — COMMENTED** [review]
I think this is likely fine but I haven't put as much thought into it as you probably have with respect to how this might break...

**Review by bdach — DISMISSED** [review]

File: osu.Game.Rulesets.Osu.Tests/TestSceneSpinnerInput.cs [hunk 1]
```diff
@@ -194,11 +196,11 @@ public void TestRewind()
             assertTotalRotation(3750, 810);
             assertTotalRotation(3500, 720);
             assertTotalRotation(3250, 530);
-            assertTotalRotation(3000, 540);
+            assertTotalRotation(3000, 450);
```
Comment (smoogipoo) [inline osu.Game.Rulesets.Osu.Tests/TestSceneSpinnerInput.cs:199]: Based on our discussions and the 3 implementations I've gone through, this is not correct. Can you explain this for me?

Up to this point, in normal playback, we've got:
1 CW spin
+ 0.5 CCW spins
+ 0.25 CW spins

Which sums up to be 1CW spin + 0.5CCW spin - the 0.25 is irrelevant because it's "returning to 0". In other words, this should be 540 as far as I can tell.

File: osu.Game.Rulesets.Osu.Tests/TestSceneSpinnerInput.cs [hunk 2]
```diff
@@ -194,11 +196,11 @@ public void TestRewind()
...
             assertTotalRotation(2750, 540);
```
Comment (smoogipoo) [inline osu.Game.Rulesets.Osu.Tests/TestSceneSpinnerInput.cs:200]: The above continues to make no sense here, where when rewinding the rotation suddenly jumps up to 540.

File: osu.Game.Rulesets.Osu.Tests/TestSceneSpinnerInput.cs [hunk 3]
```diff
@@ -194,11 +196,11 @@ public void TestRewind()
             assertTotalRotation(3750, 810);
             assertTotalRotation(3500, 720);
             assertTotalRotation(3250, 530);
```
Comment (smoogipoo) [inline osu.Game.Rulesets.Osu.Tests/TestSceneSpinnerInput.cs:198]: I'm fairly certain this is wrong as well but is saved by the test lenience.

It's hard to assert what this value should actually be since it's not explicitly defined, but... If this sequence is a 450deg delta over 500ms, then at 3250ms we have half of that (225deg). Since the rotation starts at -90deg, then this should land at `-90 + 225 = 135`. But, since we're in a segment that has reached -180deg, this is still dominated by `360 + 180 = 540`.

So I believe the correct assertion here is `540 + 135 = 675`

Comment (smoogipoo) [reply osu.Game.Rulesets.Osu.Tests/TestSceneSpinnerInput.cs:198]: This is primarily pointing out that, Strictly speaking, _all of these_ should be decreasing as time is rewound.

**Review by smoogipoo — CHANGES_REQUESTED** [review]

File: osu.Game.Rulesets.Osu.Tests/TestSceneSpinnerInput.cs (hunk 1 above)
Comment (peppy) [reply osu.Game.Rulesets.Osu.Tests/TestSceneSpinnerInput.cs:199]: I'll go through the test discrepancies, but on a quick check you're likely right. I mentioned in the OP that i didn't really go over tests with a fine comb, just made the ones you wrote pass.

That said, I don't think this is an issue with the algorithm. Ignoring the test failures, It'd be helpful to know if you're on board with the algorithm itself so I can either prioritise fixing the tests to get it in for this release, or hold it back to give us more time to discuss.

Comment (smoogipoo) [reply osu.Game.Rulesets.Osu.Tests/TestSceneSpinnerInput.cs:199]: Before I go into the algorithm, have you deviated from the original goal of requiring crossing 0 that resulted in these tests being the way they are now?

Comment (peppy) [reply osu.Game.Rulesets.Osu.Tests/TestSceneSpinnerInput.cs:199]: No, I don't believe so, that part should still be the same.

Most of the change is in the handling of rewind (simplifying it).

File: osu.Game.Rulesets.Osu/Objects/Drawables/SpinnerSpinHistory.cs [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+        public float TotalRotation => 360 * completedSpins.Count + currentSpinMaxRotation;
+
+        private readonly Stack<CompletedSpin> completedSpins = new Stack<CompletedSpin>();
+
+        /// <summary>
+        /// The total accumulated rotation.
+        /// </summary>
+        private float totalAbsoluteRotation;
+
+        private float totalAbsoluteRotationAtLastCompletion;
```
Comment (smoogipoo) [inline osu.Game.Rulesets.Osu/Objects/Drawables/SpinnerSpinHistory.cs:38]: So this isn't "absolute" rotation, it's just accumulated rotation? Maybe rename `totalAccumulatedRotation`/`totalAccumulatedRotationAtLastCompletion`?

**Comment by smoogipoo** [comment]
I don't think this can ever handle anything that happens within one "spin" (as I think you've termed it). I'm fairly certain rewinding is completely broken and irrecoverable with this implementation.

File: osu.Game.Rulesets.Osu.Tests/SpinnerSpinHistoryTest.cs [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+        public void TestRewindIntoSegmentThatHasNotCrossedZero()
+        {
+            history.ReportDelta(1000, -180);
+            Assert.That(history.TotalRotation, Is.EqualTo(180));
+            history.ReportDelta(1500, 90);
+            Assert.That(history.TotalRotation, Is.EqualTo(180));
+            history.ReportDelta(2000, 450);
+            Assert.That(history.TotalRotation, Is.EqualTo(360));
+            history.ReportDelta(1750, -45);
+            Assert.That(history.TotalRotation, Is.EqualTo(315));
```
Comment (smoogipoo) [inline osu.Game.Rulesets.Osu.Tests/SpinnerSpinHistoryTest.cs:131]: The original test was correct - this should be 180 because we've rewound somewhere in the segment `[1000, 2000)`. We haven't done (and aren't in the process of doing) the 450 rotation yet.

```suggestion
            history.ReportDelta(1750, -45);
            Assert.That(history.TotalRotation, Is.EqualTo(180));
```

**Review by smoogipoo — CHANGES_REQUESTED** [review]

**Comment by peppy** [comment]
> I don't think this can ever handle anything that happens within one "spin" (as I think you've termed it). I'm fairly certain rewinding is completely broken and irrecoverable with this implementation.

Did you want to wait to discuss this IRL, or did you have a specific test showing where you think this is broken?

File: osu.Game.Rulesets.Osu.Tests/SpinnerSpinHistoryTest.cs (hunk 5 above)
Comment (peppy) [reply osu.Game.Rulesets.Osu.Tests/SpinnerSpinHistoryTest.cs:131]: The change is intentional and correct to my eyes, but explaining this may require a voice call if it doesn't immediately make sense as to why.

File: osu.Game.Rulesets.Osu/Objects/Drawables/SpinnerSpinHistory.cs (hunk 4 above)
Comment (peppy) [reply osu.Game.Rulesets.Osu/Objects/Drawables/SpinnerSpinHistory.cs:38]: Yep, can go with that. I was having a hard time naming this because it's both accumulated and absolute.

File: osu.Game.Rulesets.Osu/Objects/Drawables/SpinnerSpinHistory.cs [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+                while (completedSpins.TryPeek(out var segment) && segment.CompletionTime > currentTime)
+                {
+                    completedSpins.Pop();
+                    totalAbsoluteRotationAtLastCompletion -= segment.Direction * 360;
+                }
+
+                // This is a best effort. We may not have enough data to match this 1:1, but that's okay.
+                // We know that the player is somewhere in a spin.
+                // In the worst case, this will be lower than expected, and recover in forward playback.
+                currentSpinMaxRotation = Math.Abs(currentSpinRotation);
```
Comment (smoogipoo) [inline osu.Game.Rulesets.Osu/Objects/Drawables/SpinnerSpinHistory.cs:116]: This doesn't actually recover safely as far as I can tell. To demonstrate, here's an adjustment to a test I've commented on above:

```diff
diff --git a/osu.Game.Rulesets.Osu.Tests/SpinnerSpinHistoryTest.cs b/osu.Game.Rulesets.Osu.Tests/SpinnerSpinHistoryTest.cs
index c20a7bbfd9..6452255110 100644
--- a/osu.Game.Rulesets.Osu.Tests/SpinnerSpinHistoryTest.cs
+++ b/osu.Game.Rulesets.Osu.Tests/SpinnerSpinHistoryTest.cs
@@ -129,6 +129,15 @@ public void TestRewindIntoSegmentThatHasNotCrossedZero()
             Assert.That(history.TotalRotation, Is.EqualTo(360));
             history.ReportDelta(1750, -45);
             Assert.That(history.TotalRotation, Is.EqualTo(315));
+
+            // This is equivalent to the t=2000 delta from above, just split into two segments to demonstrate the failure.
+            history.ReportDelta(1800, 50);
+            Assert.That(history.TotalRotation, Is.Not.EqualTo(365)); // We should not have crossed a full spin at this point.
+            Assert.That(history.TotalRotation, Is.EqualTo(180)); // Expected value - max accumulated rotation within one spin.
+
+            history.ReportDelta(2000, 400);
+            Assert.That(history.TotalRotation, Is.Not.EqualTo(765)); // We should not have 2 full spins.
+            Assert.That(history.TotalRotation, Is.EqualTo(360)); // Expected value - exactly one full spin.
         }

         [Test]
```

In other words, the worst case is actually that you have extra spins, not fewer spins.

**Review by smoogipoo — CHANGES_REQUESTED** [review]

File: osu.Game.Rulesets.Osu.Tests/TestSceneSpinnerInput.cs [hunk 7]
```diff
@@ -194,11 +196,11 @@ public void TestRewind()
             assertTotalRotation(3750, 810);
             assertTotalRotation(3500, 720);
```
Comment (smoogipoo) [inline osu.Game.Rulesets.Osu.Tests/TestSceneSpinnerInput.cs:197]: As discussed, it may be good to have these assert `(4000, 900)` again after each rewind step, to ensure forward playback is still correct.

**Review by smoogipoo — CHANGES_REQUESTED** [review]

File: osu.Game.Rulesets.Osu.Tests/SpinnerSpinHistoryTest.cs (hunk 5 above)
Comment (smoogipoo) [reply osu.Game.Rulesets.Osu.Tests/SpinnerSpinHistoryTest.cs:131]: As discussed in voice, the primary issue here is that it should be impossible for the (1750, -45) report to exist as it's passing through multiple 180-degree turns, even without important frames. I still feel like this test is illogical in its current existence, and either needs to be updated (as I did - passing through 0 from a point very close to 0) or removed.

**Comment by smoogipoo** [comment]
I still find it exceedingly hard to follow through what the total rotation should be given a sequence of movements, without actually seeing the final value first in practice...

This still does not make sense to me: https://github.com/ppy/osu/pull/25157#discussion_r1368148474
As does this: https://github.com/ppy/osu/pull/25157#discussion_r1368148966

But the tests works sooooo... ./shrug?

I don't know if this is an algorithmic failing (especially since the above is a visual test, so it's respecting frame stability) (i.e. the algorithm is still working but giving extremely unintuitive values) or just more time needs to be spend understanding how the algorithm leads to this, or what. Just noting it down...

File: osu.Game.Rulesets.Osu.Tests/TestSceneSpinnerInput.cs (hunk 7 above)
Comment (peppy) [reply osu.Game.Rulesets.Osu.Tests/TestSceneSpinnerInput.cs:201]: Have added this, fixed the clock interpolation and removed the test lenience as it is now precise \o/.

**Comment by smoogipoo** [comment]
Looks like the test changes broke everything else :(

**Comment by peppy** [comment]
Weird, I ran a full test run too.

**Comment by smoogipoo** [comment]
The editor behaves weirdly around spinners - sometimes showing negative RPM values, bonus score starting at high points, etc. I don't believe that's ever going to be fixable by this, correct? We would either need to support exactly the case I was trying to fix before, or somehow enable frame stability around just spinners (which sounds like terrible UX for the editor).

**Review by smoogipoo — APPROVED** [review]
If this is to be merged, it is only on the basis that rewinding is a TBD/rudimentary implementation. I'm still not happy with it and I can guarantee that it's going to give us a lot of headaches in its current form - the editor as I've mentioned is one, but also potentially instant-retry as I've implemented [in the past](https://github.com/smoogipoo/osu/tree/instant-retry), and anything else that may rely on non-frame-stable playback.

**Comment by peppy** [comment]
Editor is an edge case which doesn't matter at this point. It was not correct before and is not correct now.

Also, spinners aren't the only thing broken by non-frame-stable playback. The editor needs frame stable playback to some degree. We need to make it efficient enough for this to be always-on, or need to (ab)use knowledge of alive objects to enable it only for nearby past/future.

**Review by bdach — APPROVED** [review]
